win_copy: added remote and content options

jborean93 opened the pull request:
##### ISSUE TYPE
 - Feature Pull Request

##### COMPONENT NAME
win_copy

##### ANSIBLE VERSION
```
ansible 2.3.0 (win_copy-extra-params 4708fea394) last updated 2017/02/17 08:11:27 (GMT +1000)
  config file =
  configured module search path = Default w/o overrides
```

##### SUMMARY

Fixed up documentation to reference content parameter
Added support for copy files from remote host with remote_src
Added check mode
Cleaned up tests
Cleaned up module to fit with current formatting standard

jhawkesworth commented:
Sorry I missed this @jborean93 - hope you don't mind I've edited the PR title - it easier to find in all the notifications I get from github.   I will take a look at this as soon as I can.

jborean93 commented:
no worries, I knew you would get to it at some point in time :)

dagwieers commented:
Unfortunately, most of what you have done I already did in PR #20405 a month ago.

jborean93 commented:
Up to you how you want to continue with this, happy to wait until you have merged yours in/if you want to incorporate in one change. I still think the changes here are different from yours and both PR's are worthwhile.

jhawkesworth commented:
Just to say I want the stuff in #20405 **and** this PR to go in.  Need some help from the core team to figure out what to do to make the copy action plugin acceptable to everyone on #20405 though.

dagwieers commented:
I also want to have both merged. Waiting a bit more.

@jborean93 This was not criticism towards you, rather frustration that things move too slowly causing more conflicts and friction than needed.

dagwieers commented on lib/ansible/modules/windows/win_copy.py:
```diff
@@ -29,21 +29,44 @@
+        description:
+            - If False, it will search for src at originating/master machine,
+              if True it will go to the remote/target machine for the src.
+        required: False
```
Please do not add `required: False` for parameters.
This is implied and it's easier to find required parameters if only they have this flag.

jborean93 replied:
Sorry I got to get out of the habit of doing this, will fix up.

dagwieers replied:
It's not a showstopper, but it is highlighted in the documentation IIRC.
So I always make the remark.

dagwieers commented on lib/ansible/modules/windows/win_copy.py:
```diff
@@ -29,21 +29,44 @@
+            - If False, it will search for src at originating/master machine,
+              if True it will go to the remote/target machine for the src.
+        required: False
+        default: False
```
By default parameters default to $null when unset. Since this is a string, it should not default to a boolean. I think it is good practice to test for $null to find if a parameter is unset.

(As a bonus, it also uniformly allows people to unset parameters in with-loops that are not required)

jborean93 replied:
This is a Boolean flag, if set will copy the files paths of the remote hosts path and not from ansible to windows like normal. Is that ok with you?

dagwieers replied:
Ouch, I was mistaken. Sorry for that. Ignore my comment :-)
But just as a reference, also booleans can have 3 values, $true, $false and $null.
And $null is again used to indicate the parameter was not set (implicitly, or explicitly).
We ignore $null values in AnsibleParam for this exact reason.

dagwieers commented on lib/ansible/modules/windows/win_copy.py:
```diff
@@ -29,21 +29,44 @@
+              if it does not end with "/", the directory itself with all
+              contents is copied. This behavior is similar to Rsync. Needs to
+              be set if C(content) is not.
+        required: False
```
Same, please remove.

jborean93 replied:
Ditto, my bad

jborean93 committed:
win_copy: added remote and content options

jborean93 commented:
@dagwieers I've just pushed a new commit that;
- merged with your changes that was just committed to devel
- fixed up some of the py documentation your outlined
- Added some tests to cover your PR as I saw we didn't have any for it.

Please let me know if you have any other changes you wish me to make :)

ready_for_review

dagwieers commented on lib/ansible/modules/windows/win_copy.ps1:
```diff
@@ -24,77 +24,132 @@ $check_mode = Get-AnsibleParam -obj $params -name "_ansible_check_mode" -type "b
-# original_basename gets set if src and dest are dirs
-# but includes subdir if the source folder contains sub folders
-# e.g. you could get subdir/foo.txt
-
```
Where did this explanation go to ? I would prefer to keep it.

jborean93 replied:
Readded it, thanks for the pickup.

dagwieers commented:
Looking good.

IMO We are doing too many checksum operations (and these can be quite expensive). So I would like to improve that as well (at some point). Instead of 5 operations, we should end up with only 3. One for the src file, one for the original dest, and for the copied src file.

This means that we expect the action plugin to transfer these checksums as part of internal parameters to avoid having to do them again in the module).  To verify the transfer was successful, unless we consider that the transfer is a trusted action, in that case we don't need any additional checksums in the module.

jborean93 committed:
readded comment about original_basename accidentally removed

nitzmahone commented:
Same deal here- gonna merge this as-is for module cutoff. There are a few little things I might go back and twiddle/tighten up at some point, but it's great to have the features and additional test coverage. Once again, many thanks to @jborean93 and everyone for the review/feedback!
